### PR TITLE
Add rotary position embeddings

### PR DESCRIPTION
**Pull request by gkielian**
This adds rotary position embedding options.

There are two options provided:

1. RoPE (conventional rope)
2. ShortRope -- a rope of shorter length :smiley: , which is more hardware friendly

Parameters are provided to sweep or perform hyperparameter RL on various 'shortrope' lengths.

### CODE REVIEW

**Review by advancedengineering — COMMENTED** [review]
1 possible problem, everything else looks great

File: model.py [hunk 1]
```diff
@@ -149,6 +149,91 @@ def forward(self, inputs):
...
         return numerator / denominator
 
+class RotaryEmbedding(nn.Module):
+
+    def __init__(self, config):
+        super().__init__()
+        self.dim = config.n_embd
+
+        # Register frequencies directly as buffers
+        self.register_buffer('freq_left', 1.0 / (10000 ** (torch.arange(0, self.dim//2).float() / self.dim//2)))
```
[the remark this answers is not in the document]
Comment (gkielian) [reply model.py:159]: Thanks for catching this! : D

Hoping this improves the perplexity from rope embedding training.

[the remark this answers is not in the document]
Comment (gkielian) [reply model.py]: Just implemented the changes, running again to see once more how this compares with regular absolute positional encoding. 

Come to think of it, I should add the bash script for testing for visibility (a good opportunity for sharing how to test sweeps of features with the repo : )

**Comment by gkielian** [comment]
image


Seems interestingly enough that at the early stages, rope does better (at least on shakespeare with characters) and later on though absolute position encodings do better. It's almost like the model needs more position info at the start of training, but eventually learns how to leverage this from abs_pos_encodings a little better.

**Comment by gkielian** [comment]
@advancedengineering Thanks again! I implemented corrections for both rope and shortrope, just shared a script that when run should create tensorboard graphs for training on shakespear_char.
Just let me know (via merging pull request) if it looks ready : )

**Pull request merged**